runners/flink/1.15/src/main/java/org/apache/beam/runners/flink/transform/sql/StatementOnlyFlinkSqlTransformTranslator.java

```diff
@@ -0,0 +1,99 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.flink.transform.sql;
+
+import com.google.auto.service.AutoService;
+import java.util.Map;
+import java.util.StringJoiner;
+import org.apache.beam.runners.core.construction.PTransformTranslation;
+import org.apache.beam.runners.core.construction.TransformPayloadTranslatorRegistrar;
+import org.apache.beam.runners.flink.FlinkCustomTransformTranslatorRegistrar;
+import org.apache.beam.runners.flink.FlinkStreamingPipelineTranslator;
+import org.apache.beam.runners.flink.FlinkStreamingTranslationContext;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.flink.table.api.bridge.java.StreamStatementSet;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class StatementOnlyFlinkSqlTransformTranslator
+    extends FlinkStreamingPipelineTranslator.StreamTransformTranslator<PTransform<PBegin, PDone>>{
+  private static final Logger LOG = LoggerFactory.getLogger(StatementOnlyFlinkSqlTransformTranslator.class);
+  public static final String FLINK_STATEMENT_ONLY_SQL_URN = "beam:transform:flink:sql-statements-only:v1";
+  private static final String INSERT_INTO = "INSERT INTO";
+
+  @Override
+  public void translateNode(PTransform<PBegin, PDone> transform, FlinkStreamingTranslationContext context) {
+    StatementOnlySqlTransform sqlTransform = (StatementOnlySqlTransform) transform;
+
+    StreamTableEnvironment tEnv = StreamTableEnvironment.create(context.getExecutionEnvironment());
+    sqlTransform.getCatalogs().forEach(tEnv::registerCatalog);
+    StringJoiner combinedStatements = new StringJoiner("\n\n");
+    StreamStatementSet ss = tEnv.createStatementSet();
+    for (String statement : sqlTransform.getStatements()) {
+      combinedStatements.add(statement);
+      if (statement.substring(0, INSERT_INTO.length()).toUpperCase().startsWith(INSERT_INTO)) {
+        ss.addInsertSql(statement);
+      } else {
+        // Not an insert into statement. Treat it as a DDL.
+        tEnv.executeSql(statement);
```

Comment:
Can there be valid normal query statements without `INSERT INTO`? I am not very sure how FLink SQL statements look like today. Maybe everything has INSERT INTO in the beginning?

Reply:
A query logic in SQL is always a `SELECT` statement. In SQL the query logic has to be represented by some entity, and such entity would either be a `Table` or a `View`. So any SELECT statement will either be a part of `CREATE TABLE` statement or `CREATE VIEW` statement. `INSERT INTO` is effectively binding a SELECT statement with an existing table instead of creating a new one.

From execution perspective, only `INSERT INTO` will trigger an action to run the query. Without an `INSERT INTO` statement, all the queries will only generate temporary views or tables. That is why we can call `tEnv.executeSql()` for all the statements other than `INSERT INTO`, because they will not really trigger an execution. And we have to append all the `INSERT INTO` statement in a `StatementSet` so that multiple `INSERT INTO` can be executed in the same Flink job.